client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/protocol/PutOutputStream.java

```diff
@@ -0,0 +1,113 @@
+package org.infinispan.client.hotrod.impl.protocol;
+
+import static org.infinispan.client.hotrod.impl.Util.await;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiFunction;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+
+public class PutOutputStream extends OutputStream {
+   private static final int BUFFER_SIZE = 8 * 1024;
+   private final BiFunction<ByteBuf, Boolean, CompletionStage<Void>> consumer;
+   private final ByteBufAllocator alloc;
+   private final AtomicInteger pendingWrites = new AtomicInteger();
+   private ByteBuf buf;
+   private volatile Throwable throwable;
+
+   public PutOutputStream(BiFunction<ByteBuf, Boolean, CompletionStage<Void>> consumer, ByteBufAllocator alloc) {
+      this.consumer = consumer;
+      this.alloc = alloc;
+   }
+
+   private void alloc() {
+      buf = alloc.buffer(BUFFER_SIZE, BUFFER_SIZE);
+   }
+
+   private void consume(ByteBuf buf, Boolean complete) {
+      var stage = consumer.apply(buf, complete)
+            .whenComplete((___, t) -> {
+               pendingWrites.decrementAndGet();
+               if (t != null) {
+                  throwable = t;
+               }
+            });
+      // Don't let more than 2 pending writes at any time
+      if (complete || pendingWrites.incrementAndGet() >= 2) {
```

Comment:
After thinking a second about this, the way pendingWrites is done actually causes added latency as it won't submit a new task until the last one completes. We should instead use a Semaphore or similar so we instead wait until the oldest one completes to return.